### PR TITLE
ACR chunked download

### PR DESCRIPTION
**Pull request by lmolkova**
Implements download portion of #30559

### CODE REVIEW

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java [hunk 1]
```diff
@@ -322,34 +317,40 @@ public Mono<DownloadBlobResult> downloadBlob(String digest) {
...
+                .addKeyValue("responseDigest", responseHeaderDigest)
+                .log(new ServiceResponseException("The digest in the response header does not match the expected digest.")));
+        }
+
+        long blobSize = getBlobSize(firstResponse.getHeaders().get(HttpHeaderName.CONTENT_RANGE));
+        List<Mono<ContainerRegistryBlobsGetChunkResponse>> others = new ArrayList<>();
+        others.add(Mono.just(firstResponse));
+        for (long p = firstResponse.getValue().getLength(); p < blobSize; p += CHUNK_SIZE) {
+            HttpRange range = new HttpRange(p, (long) CHUNK_SIZE);
+            others.add(blobsImpl.getChunkWithResponseAsync(repositoryName, digest, range.toString(), context));
```
Comment (alzimmermsft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java:350]: Does Container Registry support HTTP preconditions? If so, we should pass a `If-Match: firstResponse.ETAG_HEADER` to ensure that we fail the download if the content changes between requests.

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java [hunk 2]
```diff
@@ -322,34 +317,40 @@ public Mono<DownloadBlobResult> downloadBlob(String digest) {
...
-            }
-        }).onErrorMap(UtilsImpl::mapException);
+        Flux<ByteBuffer> content =
+            blobsImpl.getChunkWithResponseAsync(repositoryName, digest, new HttpRange(0, (long) CHUNK_SIZE).toString(), context)
+                .flatMapMany(firstResponse -> getAllChunks(firstResponse, digest, context))
+                .flatMapSequential(chunk -> chunk.getValue().toFluxByteBuffer(), 1);
+        return Mono.just(ConstructorAccessors.createBlobDownloadResult(digest, content));
+    }
+
+    private Flux<ContainerRegistryBlobsGetChunkResponse> getAllChunks(ContainerRegistryBlobsGetChunkResponse firstResponse, String digest, Context context) {
```
Comment (alzimmermsft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java:336]: I'll leave this as an open question. 

This design is a bit different from `azure-storage-blob` where chunked download is only leveraged when performing download to file operations as disk writing should support non-sequential writes. Or, better put if we concurrently download ranges 0-4MB and 12MB-16MB, and before 4-8MB and 8MB-12MB complete, this is fine as the file ranges can still be written correctly.

This however is a case of we're downloading an unknown size container, which in this case `azure-storage-blob` uses a concept of open-ended download, `Range: start-` which indicates download from a given start point until completion. The download is then tracked for progression, such as 12MB has been downloaded, and resuming logic is added on top if there is a failure, so if at 12MB a network error happens another download call is made with `Range: progress-` to resume the download until either completion or exhaustion of retries (such as a complete networking failure).

To me this should be following the latter logic.

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ConstructorAccessors.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+    public static void setBlobDownloadResultAccessor(final BlobDownloadAsyncResultConstructorAccessor accessor) {
+        blobDownloadAccessor = accessor;
+    }
+
+    public static BlobDownloadAsyncResult createBlobDownloadResult(String digest, Flux<ByteBuffer> content) {
+        if (blobDownloadAccessor == null) {
+            try {
+                // it's possible that nobody yet created BlobDownloadAsyncResult, so we'll need to force its static section to run and set accessor.
+                Class.forName(BlobDownloadAsyncResult.class.getName(), true, BlobDownloadAsyncResultConstructorAccessor.class.getClassLoader());
```
[the remark this answers is not in the document]
Comment (lmolkova) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/ConstructorAccessors.java:28]: I want to keep `BlobDownloadAsyncResult` constructor package-private

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java (hunk 1 above)
Comment (lmolkova) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java:350]: the content is immutable  - the digest is sha256 of the content and it's used to get the content. We also validate it at the end.

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java (hunk 2 above)
Comment (lmolkova) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java:336]: yes, thanks for pointing this out and explaining how it works. I'd like to do it after finalizing the feature work and API changes.  I created #33382 to track it.

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java [hunk 4]
```diff
@@ -73,7 +75,7 @@ public final class UtilsImpl {
     public static final HttpHeaderName DOCKER_DIGEST_HEADER_NAME = HttpHeaderName.fromString("docker-content-digest");
     public static final String OCI_MANIFEST_MEDIA_TYPE = "application/vnd.oci.image.manifest.v1+json";
     public static final String CONTAINER_REGISTRY_TRACING_NAMESPACE_VALUE = "Microsoft.ContainerRegistry";
-
+    public static final int CHUNK_SIZE = 4 * 1024 * 1024;
```
[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java:78]: I was wrong on this, this is only done in the case of upload.

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java [hunk 5]
```diff
@@ -185,8 +187,29 @@ public static String computeDigest(ByteBuffer buffer) {
...
+            throw LOGGER.atError()
+                .addKeyValue("requestedDigest", requestedDigest)
+                .addKeyValue("actualDigest", () -> "sha256:" + sha256)
+                .log(new ServiceResponseException("The digest in the response does not match the expected digest."));
+        }
+    }
+
     private static final char[] HEX_ARRAY = "0123456789abcdef".toCharArray();
-    private static String byteArrayToHex(byte[] bytes) {
+    public static String byteArrayToHex(byte[] bytes) {
```
Comment (alzimmermsft) [inline sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java:212]: We use this is a handful of places throughout the SDKs, mind filing an issue to make this a utility method in CoreUtils

**Review by alzimmermsft — APPROVED** [review]

File: sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java (hunk 5 above)
Comment (lmolkova) [reply sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/implementation/UtilsImpl.java:212]: sure, #33405

**Pull request merged**